### ISSUE TITLE
User and Role specifications.

### ISSUE THREAD

**Issue by kotp** [marcric/roodle_lms#22]
Building the User and Role specifications before we write the program.

**Comment by marcric** [comment]
Please provide a short step by step narrative of what are you doing and how we can follow your steps.

**Comment by kotp** [comment]
Basically, think about what behavior you want, from your controller or model.  Write the specification, thinking about the context.  So, for example, to Describe a User, what should the user be able to do, what should it not be able to do, and what should it be able to do or not do in what context.

I took all of the features mentioned in the group discussion, made them pending specifications.  It is now our job to determine what each specification means, write those out where we have "Pending" in the specifications, and get the specification to fail, then write the model behaviors to match.

**Comment by marcric** [comment]
Ok,

I can see that here: https://github.com/marcric/roodle_lms/blob/master/spec/models/user_spec.rb

But I have just pulled out the project, run "bundle exec rspec", and get nothing useful:

http://twitpic.com/3jxi0h

Do I miss something?

**Comment by kotp** [comment]
rake spec
or 
bundle exec rspec -fd spec
The -fd is optional, in case you want to have the full report.

**Comment by marcric** [comment]
OK, much better now:

http://twitpic.com/3jycrs

http://twitpic.com/3jycxx

http://twitpic.com/3jyd1j

Should we improve the Test App Wiki page with that?

**Comment by kotp** [comment]
Updated [Testing the Application](https://github.com/marcric/roodle_lms/wiki/Testing-the-Application)

**Comment by marcric** [comment]
That is still strange...

I'm at work now, and couldn't test it, but supposing the "-fd" options are not necessary, the need of using this one:

bundle exec rspec "./spec/models/user_spec.rb"

Is even more verbose, and appears to me as a missing configuration.

The command "bundle exec rspec" should be sufficient on any platform.

Don't you think?

**Comment by kotp** [comment]
Don't forget to do:

```
bundle install
```

and then

```
bundle exec rspec spec
```

**Comment by marcric** [comment]
E:_AA-Roodle\roodle_lms>bundle check
The Gemfile's dependencies are satisfied

E:_AA-Roodle\roodle_lms>bundle exec rspec
No examples were matched. Perhaps {:if=>#<Proc:0x177a518@D:/Ruby192/lib/ruby/gems/1.9.1/gems/rspec-core-2.3.0/lib/rspec/core/configuration.rb:50 (lambda)>, :unless=>#<Proc:0x177a500@D:/Ruby192/lib/ruby/gems/1.9.1/gems/rspec-core-2.3.0/lib/rspec/core/configuration.rb:51 (lambda)>} is excluding everything?

Finished in 0 seconds
0 examples, 0 failures

E:_AA-Roodle\roodle_lms>

**Comment by kotp** [comment]
```
bundle exec rspec spec
```

Or better yet, let Rake worry about the details unless you need specific options:

```
rake spec
```

**Comment by marcric** [comment]
OK, they both works fine in the same way.

So, 

```
bundle exec rspec spec
```

Will be preferable, by explicitly using "bundle".

**Comment by kotp** [comment]
Probably.  I think though that you shouldn't forget the rake tasks that are available.  Because anything that ends up being commonly done, with regards to the tests, that we may need to customize, can be done through the rake tasks.

For example, having a rake task that provides html output to the doc directory (which can be ignored in this project) may give you nice output reports in a browser, and we can do that by doing rake spec:html (if we set that task up.)

**Comment by marcric** [comment]
I have used the html output facility in 2009 in some Ruby Dojo sessions, it is nice and we can use it.

We just need to provide step by step documentation on our Wiki pages to anyone interested run the tests and get the same expected results.

We should use the tools, facilities and options we agree the best at the time being, and document it's usage the best we can here.

Rake for instance has a ton of options with specific documentation. We just need to document here the options we have used briefly.

"We have used the "rake spec:html" option to generate a nice output report".

So any user will understand how we get the final product and reproduce the steps for his needs.

It is an LMS project that teaches by itself .
